source_code/main_mcu/src/fido2/fido2.c

```diff
@@ -0,0 +1,493 @@
+*   \param  hash of client data
+*   \param  out buffer with signature
+*   \return void
+*/
+static void fido2_calc_attestation_signature(uint8_t const *data, int datalen, uint8_t const * client_data_hash, uint8_t * sigbuf)
+{
+    #define HASH_LEN 32
+    uint8_t hashbuf[HASH_LEN];
+
+    crypto_sha256_init();
+    crypto_sha256_update(data, datalen);
+    crypto_sha256_update(client_data_hash, CLIENT_DATA_HASH_LEN);
+    crypto_sha256_final(hashbuf);
+
+    crypto_ecc256_sign(hashbuf, HASH_LEN, sigbuf);
+}
+
+/*! \fn     fido2_cbor_encode_public_key(uint8_t *buf, uint32_t bufLen, ecc256_pub_key const *pub_key)
+*   \brief  CBOR encoder for the public key. All cbor encoding and decoding belongs to aux_mcu.
+*           Unfortunately, to be able to create attestation signature we have to encode the public key
+*           in cbor format so we have to do some cbor encoding on main_mcu also.
+*           Essentiallyt the same as Solo function ctap_add_coseKey() but
+*           removed error checking and pass in buffer instead of the cbor encoder object
+*           in addition to hardcoding algtype.
+*   \param  buffer for encoded output
+*   \param  buffer length
+*   \param  public key
+*   \return encoded public key buffer size
+*/
+static uint32_t fido2_cbor_encode_public_key(uint8_t *buf, uint32_t bufLen, ecc256_pub_key const *pub_key)
+{
+    CborEncoder cose_key;
+    CborEncoder map;
+
+    cbor_encoder_init(&cose_key, buf, bufLen, 0);
+
+    cbor_encoder_create_map(&cose_key, &map, 5);
+
+    cbor_encode_int(&map, COSE_KEY_LABEL_KTY);
+    cbor_encode_int(&map, COSE_KEY_KTY_EC2);
+
+    cbor_encode_int(&map, COSE_KEY_LABEL_ALG);
+    cbor_encode_int(&map, COSE_ALG_ES256);
+
+    cbor_encode_int(&map, COSE_KEY_LABEL_CRV);
+    cbor_encode_int(&map, COSE_KEY_CRV_P256);
+
+    cbor_encode_int(&map, COSE_KEY_LABEL_X);
+    cbor_encode_byte_string(&map, pub_key->x, 32);
+    cbor_encode_int(&map, COSE_KEY_LABEL_Y);
+    cbor_encode_byte_string(&map, pub_key->y, 32);
+    cbor_encoder_close_container(&cose_key, &map);
+
+    return cbor_encoder_get_buffer_size(&cose_key, buf);
+}
+
+/*! \fn     fido2_process_exclude_list_item(FIDO2_auth_cred_req_message_t const *request, FIDO2_auth_cred_rsp_message_t *response)
+*   \brief  Process Exclude list check from aux_mcu.
+*           Checks if tag already exists. Returns 1 if credential exists or 0
+*           otherwise. If tag exists prompt the user and wait for user ack.
+*   \param  incoming messsage request
+*   \param  outgoing response
+*   \return void
+*/
+void fido2_process_exclude_list_item(FIDO2_auth_cred_req_message_t const *request, FIDO2_auth_cred_rsp_message_t *response)
+{
+    credential_DB_rec_t *cred_rec = fido2_get_credential(request->rpID, request->cred_ID.tag);
+    bool found= (cred_rec != NULL) ? 1 : 0;
+
+    memset(response, 0, sizeof(*response));
+
+    if (found)
+    {
+        fido2_prompt_user("Credential exists already and in exclude list", true, 10);
```

Comment:
pretty sure the user wouldn't understand anything if he'd see that on the display :D
does this need to be displayed in the first place?

Reply:
I know. That is a placeholder. We need a message telling the user an error occurred. Feel free to create a good message when you implement the display functionality,

Reply:
ok!